drivers/pwm/pwm_led_esp32.c

```diff
@@ -0,0 +1,470 @@
+	set_mask32(tick_sel << LEDC_TICK_SEL_LSTIMER0_S, timer_addr);
+	set_mask32(bit_num & LEDC_LSTIMER0_LIM_M, timer_addr);
+
+	if (speed_mode) {
+		/* update div_num and bit_num */
+		sys_set_bit(timer_addr, LEDC_LSTIMER0_PARA_UP_S);
+	}
+
+	/* reset low speed timer */
+	sys_set_bit(timer_addr, LEDC_LSTIMER0_RST_S);
+	sys_clear_bit(timer_addr, LEDC_LSTIMER0_RST_S);
+
+	return 0;
+}
+
+/* period_cycles is not used, set frequency on menuconfig instead. */
+static int pwm_led_esp32_pin_set_cycles(struct device *dev,
+					u32_t pwm, u32_t period_cycles,
+					u32_t pulse_cycles)
+{
+	ARG_UNUSED(period_cycles);
+
+	int speed_mode;
+	int ret;
+
+	const struct pwm_led_esp32_config * const config =
+		(struct pwm_led_esp32_config *) dev->config->config_info;
+
+	struct pwm_led_esp32_data *data = dev->driver_data;
+
+	if (pwm >= config->gpio_range_hs.lo &&
+	    pwm <= config->gpio_range_hs.hi) {
+		speed_mode = PWM_LED_ESP32_HIGH_SPEED;
+	} else if (pwm >= config->gpio_range_ls.lo &&
+		   pwm <= config->gpio_range_ls.hi) {
+		speed_mode = PWM_LED_ESP32_LOW_SPEED;
+	} else {
+		return -EINVAL;
+	}
+
+	if (data->pin_ch[pwm] == PWM_LED_ESP32_CHANNEL_NOT_USED) {
+		if (data->hs_ch_used >= 8 || data->ls_ch_used >= 8) {
+			return -EINVAL;
+		}
+		if (speed_mode == PWM_LED_ESP32_HIGH_SPEED) {
+			data->pin_ch[pwm] = data->hs_ch_used;
+			data->hs_ch_used++;
+		} else {
+			data->pin_ch[pwm] = data->ls_ch_used;
+			data->ls_ch_used++;
+		}
+	}
+
+	if (period_cycles == 0 || pulse_cycles > period_cycles) {
+		return -EINVAL;
+	}
+
+	esp32_enable_peripheral(&config->peripheral);
+
+	/* Set timer */
+	ret = pwm_led_esp32_timer_set(speed_mode,
+			config->timer_cfg[speed_mode].timer_sel,
+			config->timer_cfg[speed_mode].bit_num,
+			config->timer_cfg[speed_mode].freq);
+	if (ret < 0) {
+		return ret;
+	}
+
+	/* Set channel */
+	ret = pwm_led_esp32_channel_set(pwm, speed_mode,
+			data->pin_ch[pwm], 0,
+			config->timer_cfg[speed_mode].timer_sel);
+	if (ret < 0) {
+		return ret;
+	}
+
+	pwm_led_esp32_duty_set(speed_mode, data->pin_ch[pwm], pulse_cycles);
+	pwm_led_esp32_update_duty(speed_mode, data->pin_ch[pwm]);
+
+	return ret;
+}
+
+static int pwm_led_esp32_get_cycles_per_sec(struct device *dev, u32_t pwm,
+					    u64_t *cycles)
+{
+	const struct pwm_led_esp32_config *config;
+
+	config = (struct pwm_led_esp32_config *) dev->config->config_info;
+
+	if (pwm >= config->gpio_range_hs.lo &&
+	    pwm <= config->gpio_range_hs.hi) {
+		*cycles = config->timer_cfg[PWM_LED_ESP32_HIGH_SPEED].freq;
+	} else if (pwm >= config->gpio_range_ls.lo &&
+		   pwm <= config->gpio_range_ls.hi) {
+		*cycles = config->timer_cfg[PWM_LED_ESP32_LOW_SPEED].freq;
+	} else {
+		return -EINVAL;
+	}
+
+	return 0;
+}
+
+static const struct pwm_driver_api pwm_led_esp32_api = {
+	.pin_set = pwm_led_esp32_pin_set_cycles,
+	.get_cycles_per_sec = pwm_led_esp32_get_cycles_per_sec,
+};
+
+/* Initialization for PWM_LED_ESP32 */
+#if defined(CONFIG_PWM_LED_ESP32_DEV_NAME_0)
```

Comment:
Is this ifdef required? This file will only be compiled if the driver has been enabled.

Reply:
I'll fix it.